File: 2024/March/3-23 143 Reorder List/ReorderList.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        listCount = 0
        stack = []
        node = head
        while node:
            stack.append(node)
            node = node.next
            listCount += 1
        
        if listCount <= 2:
            return
        
        for i in range(0, listCount, 2):
            temp = head.next
            head.next = stack.pop()
            head.next.next = temp
            head = head.next.next
        head.next = None
        

File: 2024/March/3-23 143 Reorder List/ReorderList.java
import java.util.Stack;

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public void reorderList(ListNode head) {
        int listCount = 0;
        Stack<ListNode> stack = new Stack<ListNode>();
        ListNode node = head;
        while(node != null) {
            stack.push(node);
            node = node.next;
            listCount++;
        }

        if(listCount <= 2) {
            return;
        }

        ListNode temp;
        for(int i = 0; i < listCount; i += 2) {
            temp = head.next;
            head.next = stack.pop();
            head.next.next = temp;
            head = head.next.next;
        }
        head.next = null;
    }
}